### ISSUE TITLE
Incorrect `num_warmup_steps` for `lr_scheduler` for multi-gpu training

### ISSUE THREAD

**Issue by cyk1337** [huggingface/accelerate#662]
### System Info

```Shell
- `Accelerate` version: 0.10.0
- Platform: Linux-3.10.0_3-0-0-12-x86_64-with-centos-6.3-Final
- Python version: 3.7.12
- Numpy version: 1.21.6
- PyTorch version (GPU?): 1.7.1 (True)
- `Accelerate` default config:
        - compute_environment: LOCAL_MACHINE
        - distributed_type: MULTI_GPU
        - mixed_precision: no
        - use_cpu: False
        - num_processes: 8
        - machine_rank: 0
        - num_machines: 1
        - main_process_ip: None
        - main_process_port: None
        - main_training_function: main
        - deepspeed_config: {}
        - fsdp_config: {}
```

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [ ] My own task or dataset (give details below)

### Reproduction

[https://github.com/huggingface/transformers/blob/f2fbe4475386bfcfb3b83d0a3223ba216a3c3a91/examples/pytorch/translation/run_translation_no_trainer.py#L533](https://github.com/huggingface/transformers/blob/f2fbe4475386bfcfb3b83d0a3223ba216a3c3a91/examples/pytorch/translation/run_translation_no_trainer.py#L533)
```bash
# define lr scheduler
lr_scheduler = get_scheduler(
        name="linear",
        optimizer=optimizer,
        num_warmup_steps=args.warmup_steps,
        num_training_steps=args.max_train_steps,
    )

...

if step % args.gradient_accumulation_steps == 0:                    
      optimizer.step()
      lr_scheduler.step() # update lr scheduler every `gradient_accumulation_steps`
      optimizer.zero_grad()
```

### Expected behavior

Is the `accelerate` consider the num of processes for `num_warmup_steps`? 
Suppose we set `args.warmup_steps=80` and train on a single 8-gpu machine, the linear learning rate will peak at 10 (i.e., `80/8`) rather than expected 80.

**Comment by cyk1337** [comment]
According to the design of `accelerate` [https://github.com/huggingface/accelerate/blob/d0f5f4a630bda69dcf89cc6d55f93c71f2af7a0d/src/accelerate/scheduler.py#L70](https://github.com/huggingface/accelerate/blob/d0f5f4a630bda69dcf89cc6d55f93c71f2af7a0d/src/accelerate/scheduler.py#L70), is it correct to set  the warmup_steps  as `warmup_steps*num_processes`?

**Comment by pacman100** [comment]
Hello @cyk1337, when you use Multi-GPU setup, the `max_train_steps` decrease by `num_gups` , i.e., `result_max_train_steps = args.max_train_steps // num_gups`. As such, by design scheduler steps too will be reduced by `num_gups` so that its LR gets to 0 when it reaches `result_max_train_steps`. Therefore, `warmup_steps` too should decrease by `num_gups`. Im your example, with multi-gpu 8 and  args.warmup_steps=80, if the warmup_steps don't decrease to 10, the number of samples it takes to get to full LR would be `80*8*batch_size_per_gpu` instead of `80*batch_size_per_gpu` if you were using single GPU. SO, in terms of samples consumed to get to full LR, reducing `warmup_steps` by `num_gpus` is required.

**Comment by cyk1337** [comment]
> when you use Multi-GPU setup, the max_train_steps decrease by num_gups , i.e., `resulting_max_train_steps = args.max_train_steps // num_gups`

Hi @pacman100, thank you for answering my question. Could you link me to `args.max_train_steps // num_gpus`? Where does it do? 

Also, without preparing `lr_scheduler`, the `warmup_steps` will not be divided by `num_gpus`.

**Comment by pacman100** [comment]
> According to the design of `accelerate`
> 
> https://github.com/huggingface/accelerate/blob/d0f5f4a630bda69dcf89cc6d55f93c71f2af7a0d/src/accelerate/scheduler.py#L70
> 
> , is it correct to set the warmup_steps as `warmup_steps*num_processes`, or just do not prepare `lr_scheduler`?

Hello @cyk1337 , the link you have provided achieves `args.max_train_steps // num_gpus` because it is steping for `num_processes` per iteration, i.e., num_gpus times per iteration. 

I didn't understand what the query was in case of not preparing `lr_scheduler`. As per the original question, it is logical to have warmup steps to be reduced in a multi-device scenario.

**Comment by cyk1337** [comment]
> when you use Multi-GPU setup, the `max_train_steps` decrease by num_gpus , i.e., `resulting_max_train_steps = args.max_train_steps // num_gpus`. As such, by design scheduler steps too will be reduced by `num_gpus` so that its LR gets to 0 when it reaches `resulting_max_train_steps`. 

Hi @pacman100, as aforementioned, suppose the `args.max_train_steps` and `warmup_steps` both divided by the `num_gpus`, the total training step (see the example [https://github.com/huggingface/transformers/blob/f2fbe4475386bfcfb3b83d0a3223ba216a3c3a91/examples/pytorch/translation/run_translation_no_trainer.py#L649](https://github.com/huggingface/transformers/blob/f2fbe4475386bfcfb3b83d0a3223ba216a3c3a91/examples/pytorch/translation/run_translation_no_trainer.py#L649)) should be `args.max_train_steps//num_gpus` rather than `args.max_train_steps`.  

Given the same training token budgets, do you mean that followings are equivalent? 
- 1 node: max_train_steps=1000, warmup_steps=10
- 2 node: max_train_steps=500 (1000//2), warmup_steps=5 (10//2).

**Comment by cyk1337** [comment]
> the link you have provided achieves args.max_train_steps // num_gpus because it is steping for num_processes per iteration, i.e., num_gpus times per iteration.

The [link](https://github.com/huggingface/accelerate/blob/d0f5f4a630bda69dcf89cc6d55f93c71f2af7a0d/src/accelerate/scheduler.py#L70) reduces the lr scheduler steps for warmup, rather than `max_train_steps`. Where was the `max_train_steps` reduced?

**Comment by cyk1337** [comment]
> what the query was in case of not preparing lr_scheduler?

Multi-gpus expand the actual batch size by `num_gpu` times, so we would like to treat the `num_gpus` batches (each gpu processes one batch) as one batch of the global steps. However, with `accelerate.prepare`, the lr scheduler steps `num_gpus` steps rather than one step. I was confused about this. Unsure whether there are misunderstandings.

**Comment by pacman100** [comment]
> Given the same training token budgets, do you mean that followings are equivalent?
> 
> * 1 node: max_train_steps=1000, warmup_steps=10
> * 2 node: max_train_steps=500 (1000//2), warmup_steps=5 (10//2).

@cyk1337 , yes. 

> Where was the max_train_steps reduced (to control the end of multi-gpu training)?

`max_train_steps` is recomputed/reduced by `num_gpus`. in the below code snippet as the length of dataloader reduces by `num_gpus` post `prepare` call.

https://github.com/huggingface/transformers/blob/f2fbe4475386bfcfb3b83d0a3223ba216a3c3a91/examples/pytorch/translation/run_translation_no_trainer.py#L545-L548

> Multi-gpus expand the actual batch size by num_gpu times, so we would like to treat the num_gpus batches (each gpu processes one batch) as one batch of the global steps. However, with accelerate.prepare, the lr scheduler steps num_gpus steps rather than one step. I was confused about this. Unsure whether there are misunderstandings.

In Multi-GPU setting, as mentioned above and in previous comments, the `max_train_steps` decrease by `num_gpus`. As such, by design, scheduler steps `num_gpus` times per iteration to account for the decrease in the  `max_train_steps`.

Hope above explanations help.

**Comment by cyk1337** [comment]
> max_train_steps is recomputed/reduced by num_gpus. in the below code snippet as the length of dataloader reduces by num_gpus post prepare call. 
[https://github.com/huggingface/transformers/blob/f2fbe4475386bfcfb3b83d0a3223ba216a3c3a91/examples/pytorch/translation/run_translation_no_trainer.py#L545-L548](https://github.com/huggingface/transformers/blob/f2fbe4475386bfcfb3b83d0a3223ba216a3c3a91/examples/pytorch/translation/run_translation_no_trainer.py#L545-L548)

Thank you for the helpful clarification.  
One more question:
`num_update_steps_per_epoch = math.ceil(len(train_dataloader) / args.gradient_accumulation_steps)`, the train_step is reduced by the `gradient_accumulation_steps` rather than the `gradient_accumulation_steps*num_gpus`. I cannot find the the `num_gpus` or`num_processes`. Could you please link me to the `num_gpus`?

**Comment by pacman100** [comment]
> max_train_steps is recomputed/reduced by num_gpus in the below code snippet as the length of dataloader reduces by num_gpus post prepare call.
>
> https://github.com/huggingface/transformers/blob/f2fbe4475386bfcfb3b83d0a3223ba216a3c3a91/examples/pytorch/translation/run_translation_no_trainer.py#L545-L548

As mentioned here, len(dataloader) decreases by num_gpu. See below code block:
https://github.com/huggingface/transformers/blob/f2fbe4475386bfcfb3b83d0a3223ba216a3c3a91/examples/pytorch/translation/run_translation_no_trainer.py#L545-L548

Before `prepare` call:
```
num_update_steps_per_epoch = math.ceil(len(train_dataloader) / args.gradient_accumulation_steps)
# let len(train_dataloader) be `orig_iters`. So above things is orig_iters/grad_acc_steps
# let num_update_steps_per_epoch be `orig_num_update_steps_per_epoch`
```

After `prepare` call:
```
num_update_steps_per_epoch = math.ceil(len(train_dataloader) / args.gradient_accumulation_steps)
# Here, len(train_dataloader) = len(self.batch_sampler) // self.num_processes
# Now len(dataloader) is len(batch_sampler) if you see here https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L696
# len(train_dataloader) = orig_iters // self.num_processes
# So, above thing is (orig_iters / self.num_processes) / args.gradient_accumulation_steps
# which is (orig_iters / args.gradient_accumulation_steps) / self.num_processes
# which is orig_num_update_steps_per_epoch / self.num_processes
```

**Comment by cyk1337** [comment]
Thank you for the detailed reply! 
So, if we use an `IterableDataset`whose `__len__` is unknown, we should manually set the `max_train_steps` as `max_train_steps//num_processes` so as to keep pace with `warmup_steps` (processed by `prepare`), correct?

**Comment by jcyk** [comment]
I encountered the same problem. 
```
for step in max_training_steps:
     optimizer.step()
     lr_scheduler.step()
```
It is very counter-intuitive for the` lr_scheduler` to step `num_processes` times at each call. I think, for most users, the primary purpose of using multiple GPUs is for a larger batch size, not for fewer training steps. And if they want to have fewer training steps, they can do it explicitly (1. just feed the reduced #training steps to the scheduler explicitly 2. if #training epochs is specified, compute the correct #training steps before feeding it to lr_scheduler). **The implicit behavior in current design is very confusing.**

**Comment by jcyk** [comment]
> Thank you for the detailed reply! So, if we use an `IterableDataset`whose `__len__` is unknown, we should manually set the `max_train_steps` as `max_train_steps//num_processes` so as to keep pace with `warmup_steps` (processed by `prepare`), correct?

and if I want to train exactly `max_train_steps` steps, I  should pass `max_train_steps * num_processes` to the `lr_scheduler` but the code should still be 
```

lr_scheduler = get_scheduler(
        ...
        num_training_steps= max_train_steps * num_processes,
    )
for step in range(max_training_steps):
     optimizer.step()
     lr_scheduler.step()
``` 

This is a very counter-intuitive  example.

**Comment by pacman100** [comment]
> I think, for most users, the primary purpose of using multiple GPUs is for a larger batch size, not for fewer training steps

The current behaviour is nothing to do with decreasing training steps. One usually does `for batch in dataloader` and in multi-gpu setting, because the batch_size increases by `num_gpus`, the number of batches decrease by `num_gups` and dataloader will run out of data if `max_train_steps` isn't adjusted to account for it. And, if you want more training steps, why not increase the number of epochs?

**Comment by pacman100** [comment]
> > Thank you for the detailed reply! So, if we use an `IterableDataset`whose `__len__` is unknown, we should manually set the `max_train_steps` as `max_train_steps//num_processes` so as to keep pace with `warmup_steps` (processed by `prepare`), correct?
> 
> and if I want to train exactly `max_train_steps` steps, I should pass `max_train_steps * num_processes` to the `lr_scheduler` but the code should still be
> 
> ```
> 
> lr_scheduler = get_scheduler(
>         ...
>         num_training_steps= max_train_steps * num_processes,
>     )
> for step in range(max_training_steps):
>      optimizer.step()
>      lr_scheduler.step()
> ```
> 
> This is a very counter-intuitive example.

@jcyk , if you want to train exactly for `max_train_steps` wherein it is based on dataloader, you will run out of data as each step now takes `num_gpus` times more samples. Also, for discussions and general queries, https://discuss.huggingface.co/c/accelerate/18 forums is the place. Please raise seperate issue in case of bug/feature request.

**Comment by pacman100** [comment]
> Thank you for the detailed reply! So, if we use an `IterableDataset`whose `__len__` is unknown, we should manually set the `max_train_steps` as `max_train_steps//num_processes` so as to keep pace with `warmup_steps` (processed by `prepare`), correct?

For `IterableDataset`, yes, you are correct, but in that example that you linked in the issue, this isn't the case.

**Comment by jcyk** [comment]
Thanks for your reply! @pacman100 
my point is that it should be better to adjust `max_train_steps` explicitly before feeding it to `lr_scheduler`.
in my case, I am not using #epochs but using #steps.

**Comment by sgugger** [comment]
@jcyk Don't send your scheduler to the `accelerator.prepare` method if you don't want it adapted to the number of steps. It's as simple as that.

**Comment by cyk1337** [comment]
> Don't send your scheduler to the accelerator.prepare method if you don't want it adapted to the number of steps. It's as simple as that.

Hi @sgugger, if so, how to save the optimizer/ lr_scheduler states (for resuming training) using deepspeed?
